### ISSUE TITLE
Nix treats 401/403/407 HTTP errors from binary cache as 404 and caches them

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nix#4445]
It's possible that a binary cache returns 401 due to a problem with authentication temporarily, but Nix will cache that response and fail until the disk cache has been cleared.

This was introduced in 88541569a2acded0e4999368a4724c96d6ed37e7, @edolstra do you remember by any chance why?

**Comment by edolstra** [comment]
Yes, as the comment says, S3 returns 403 for missing files in unlistable buckets.

**Comment by domenkozar** [comment]
This is probably another reason to do https://github.com/NixOS/nix/issues/4028

**Comment by domenkozar** [comment]
Note that it also treats 401 and 407 as 404.

**Comment by domenkozar** [comment]
glad there was finally motivation to fix this after 5 years in https://github.com/NixOS/nix/commit/4bd000060376fa99a188e125903bc7e04988fd1d
